### ISSUE TITLE
Consider changing `sf::Texture::invalidateMipmap()` from private to public

### ISSUE THREAD

**Issue by selendym** [SFML/SFML#2919]
### Prerequisite Checklist

- [X] I searched for [existing issues](https://github.com/search?q=repo%3ASFML%2FSFML&type=issues) to prevent duplicates
- [X] I searched for [existing discussions on the forum](https://www.google.com/search?q=site%3Ahttps%3A%2F%2Fen.sfml-dev.org) to prevent duplicates
- [X] I understand [SFML's scope](https://www.sfml-dev.org/contribute.php#general-considerations) and believe the feature fits it

### Describe your feature request here

Hello.

I'd like to suggest making [`sf::Texture::invalidateMipmap()`](https://github.com/SFML/SFML/blob/7ec3760fe8b451ef58b73df199066e6396a060f9/include/SFML/Graphics/Texture.hpp#L615) public, if possible, as it is currently private for no apparent reason, other than lack of a use case.

There is some earlier discussion about this [here](https://github.com/SFML/SFML/pull/973#issuecomment-142670725); to quote:
> I don't see any reason to make it public if we already make sure any modification by the user results in proper invalidation.

The problem with this reasoning is that even if the texture data does not change, the texture appearance does change between having mipmaps generated and not having them.
Currently, there is only a way to enable them, while disabling requires reloading the texture, which is less than optimal.
My use case for sfml is image viewing, so even slight changes in the perceived appearance are quite important.

Best regards.

### Use Cases

- image viewing
- conserving gpu memory by dropping generated mipmaps while retaining the texture
- api consistence: if something can be enabled, it should be possible to disable it too

**Comment by eXpl0it3r** [comment]
> the texture appearance does change between having mipmaps generated and not having them

Can you elaborate what "texture appearance" means?

> api consistence: if something can be enabled, it should be possible to disable it too

So is this about fully disabling the mipmaps or just invalidating the generate ones (those are two different features)?

**Comment by selendym** [comment]
> Can you elaborate what "texture appearance" means?

I mean the effect that generating mipmaps has on what is drawn on the screen.
My observation has been that without mipmaps, the result is sharper but more jagged, while with mipmaps it is blurrier but smoother.
I can add an example if needed.

> So is this about fully disabling the mipmaps or just invalidating the generate ones (those are two different features)?

To be honest, I'm not sure I understand the difference here.
What I mean is to undo (at least visually, even if resources are not freed) the effect that generating mipmaps has.
I've assumed that this is what invalidating mipmaps does, but feel free to correct me if I'm wrong.
Currently, I've achieved the above by reloading a texture (without generating mipmaps), but this feels somewhat wasteful, since the texture data itself doesn't change.

**Comment by eXpl0it3r** [comment]
> My observation has been that without mipmaps, the result is sharper but more jagged, while with mipmaps it is blurrier but smoother.

You said, you're making an image view application. I assume you're display the image at a lower resolution than the texture has?
That would explain why you get a scaled down version with generated mipmaps.

One thing that's unclear to me is, why do you generate mipmaps in the first place if it creates an undesired effect?

> To be honest, I'm not sure I understand the difference here. What I mean is to undo (at least visually, even if resources are not freed) the effect that generating mipmaps has. I've assumed that this is what invalidating mipmaps does, but feel free to correct me if I'm wrong. Currently, I've achieved the above by reloading a texture (without generating mipmaps), but this feels somewhat wasteful, since the texture data itself doesn't change.

Okay, yeah it's a subtle difference, but you just want to disable it and not having regenerated or similar (i.e. invalidation doesn't necessarily imply disabling).

**Comment by selendym** [comment]
> You said, you're making an image view application. I assume you're display the image at a lower resolution than the texture has?
> That would explain why you get a scaled down version with generated mipmaps.

Yes, it's an interactive image viewer, so zooming in and out is core functionality.
The effect mipmaps have is mostly noticeable only when displaying huge images zoomed out considerably, say, a 8000x8000 image zoomed out to fit a 1920x1080 screen.

> One thing that's unclear to me is, why do you generate mipmaps in the first place if it creates an undesired effect?

Whether or not mipmaps have an undesired effect depends on at least image content, zoom level, and user preference, so I've provided an on-off toggle key to match this.
Effectively, this makes mipmaps just another visual filter, much like the smoothing filter.
By toggling mipmaps on and off the user can choose which one to use for that particular situation, so it's not really in my power to say "this one is better".

I understand this might be too niche a use case to justify any changes in sfml.
My initial assumption was that the functionality already exists in sfml but is just private and making it public would solve the issue.
If this is not so, I have a hard time justifying anything more extensive, since this is only a slight inefficiency and reloading a texture does the job reasonably well.
